node/src/chain/stall_chain_fsm.rs

```diff
@@ -0,0 +1,302 @@
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+//
+// Copyright (c) DUSK NETWORK. All rights reserved.
+
+use node_data::{
+    ledger::{to_str, Block},
+    message::payload::Inv,
+};
+use std::{
+    collections::BTreeMap,
+    sync::Arc,
+    time::{SystemTime, UNIX_EPOCH},
+};
+use tokio::sync::RwLock;
+use tracing::{error, info, trace, warn};
+
+use crate::{
+    database::{self, Ledger},
+    vm::VMExecution,
+    Network,
+};
+
+use super::acceptor::Acceptor;
+
+const STALLED_TIMEOUT: u64 = 60; // seconds
+
+#[derive(Debug, Clone, Copy, PartialEq)]
+pub(crate) enum State {
+    /// Blocks are being accepted
+    Running,
+    /// No block has been accepted recently
+    ///
+    /// Node might be stuck on non-main branch and might need to recover
+    Stalled,
+    /// Node is disconnected from the main branch
+    StalledOnFork,
+}
+
+/// Implements a simple FSM to detect a stalled state of the chain
+pub(crate) struct StalledChainFSM<DB: database::DB, N: Network, VM: VMExecution>
+{
+    acc: Arc<RwLock<Acceptor<N, DB, VM>>>,
+
+    state: State,
+    recovery_blocks: BTreeMap<u64, Block>,
+
+    /// Latest finalized block
+    latest_finalized: Block,
+
+    /// Tip of the chain with timestamp
+    tip: (Block, u64),
+}
+
+impl<DB: database::DB, N: Network, VM: VMExecution> StalledChainFSM<DB, N, VM> {
+    pub(crate) async fn new_with_acc(
+        acc: Arc<RwLock<Acceptor<N, DB, VM>>>,
+    ) -> Self {
+        let tip = acc.read().await.get_curr_tip().await;
+        let latest_finalized = acc
+            .read()
+            .await
+            .get_latest_final_block()
+            .await
+            .expect("latest final block should exist");
+
+        let mut sm = Self {
+            state: State::Running,
+            recovery_blocks: BTreeMap::new(),
+            tip: Default::default(),
+            latest_finalized,
+            acc,
+        };
+
+        sm.update_tip(tip.inner().clone());
+        sm
+    }
+
+    /// Handles block received event
+    ///
+    /// Returns the new state of the FSM after processing the block
+    pub(crate) async fn on_block_received(&mut self, blk: &Block) -> State {
+        trace!(
+            event = "chain.block_received",
+            hash = to_str(&blk.header().hash),
+            height = blk.header().height,
+            iter = blk.header().iteration,
+        );
+
+        let tip = self.acc.read().await.get_curr_tip().await;
+
+        if self.tip.0.header().hash != tip.inner().header().hash {
+            // Tip has changed, which means a new block is accepted either due
+            // to normal block acceptance or fallback execution
+            self.recovery_blocks.clear();
+
+            if tip.is_final() {
+                self.latest_finalized = tip.inner().clone();
+            }
+
+            self.update_tip(tip.inner().clone());
+            self.state_transition(State::Running);
+        }
```

Comment:
We have to continue so to call "on_running" method

Reply:
`on_running` only does something if `tip` was updated more than `STALLED_TIMEOUT` seconds ago.
Since we just updated the tip, I don't think this will ever be possible.
Basically calling `on_running` after `update_tip` has no effect as far as I can see.